### PR TITLE
initial implementation of alias annotation

### PR DESCRIPTION
**Pull request by James201311**

### CODE REVIEW

File: encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            try {
+              method.setAccessible(true);
+              obj = method.invoke(annotation);
+            } catch (IllegalAccessException e) {
+              throw new RuntimeException(e);
+            } catch (InvocationTargetException e) {
+              throw new RuntimeException(e);
+            }
+            if (propertyName.equals("name")) {
+              String val = (String) obj;
```
Comment (vkryachko) [inline encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java:80]: This cast could fail if my annotation has a different type for the property, i.e.

```java
@Alias
@interface Foo {
  @Alias.Property("name")
  int foo();
}
```
We should throw an `EncodingException` with a descriptive message.

File: encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            if (property == null) {
+              continue;
+            }
+            String propertyName = property.value();
+            Object obj = null;
+            try {
+              method.setAccessible(true);
+              obj = method.invoke(annotation);
+            } catch (IllegalAccessException e) {
+              throw new RuntimeException(e);
```
Comment (vkryachko) [inline encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java:75]:
```suggestion
              throw new EncodingException(e);
```

File: encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            } catch (InvocationTargetException e) {
+              throw new RuntimeException(e);
```
Comment (vkryachko) [inline encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java:77]:
```suggestion
              throw new EncodingException(e);
```

File: encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      decodingKey = fieldName((Method) accObj);
+    } else {
+      throw new EncodingException("Constructor shouldn't be used to get its decoding key");
+    }
+    this.inline = false;
+    this.ignored = false;
+
+    // internal annotations
+    if (accObj.isAnnotationPresent(Encodable.Ignore.class)) {
+      this.setIgnored(true);
```
Comment (vkryachko) [inline encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java:43]: Consider removing all setters and set directly through the field.

File: encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      Encodable.Field annotation = accObj.getAnnotation(Encodable.Field.class);
+      if (annotation != null && annotation.name().length() > 0) {
+        this.setDecodingKey(annotation.name());
+      }
+      if (annotation != null) {
+        this.setInline(annotation.inline());
+      }
+    }
+
+    // alias annotations
```
Comment (vkryachko) [inline encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java:56]: looks like this only supports `Ignore` and `Field`, can you add a `TODO` to add support for custom `ExtraProperty` annotations?

**Comment by James201311** [comment]
@vkryachko Hey Vlad, is it possible to add my account to the firebase team? I am losing access to the push code.

**Comment by vkryachko** [comment]
@James201311 I am pretty sure I can't, but you should be able to fork the repo, push to it and then create PRs into this repo. This is a pretty standard practice on github.

**Comment by James201311** [comment]
Moved to https://github.com/firebase/firebase-android-sdk/pull/1913

**Pull request closed without merging**